include/deal.II/lac/linear_operator.h

```diff
@@ -0,0 +1,632 @@
+{
+public:
+
+  /**
+   * Create an empty LinearOperator object
+   */
+  LinearOperator ()
+  {
+    vmult = [](Range &, const Domain &)
+    {
+      Assert(false, dealii::ExcMessage("Uninitialized LinearOperator<Range, "
+                                       "Domain>::vmult called"));
+    };
+
+    vmult_add = [](Range &, const Domain &)
+    {
+      Assert(false, dealii::ExcMessage("Uninitialized LinearOperator<Range, "
+                                       "Domain>::vmult_add called"));
+    };
+
+    Tvmult = [](Domain &, const Range &)
+    {
+      Assert(false, dealii::ExcMessage("Uninitialized LinearOperator<Range, "
+                                       "Domain>::Tvmult called"));
+    };
+
+    Tvmult_add = [](Domain &, const Range &)
+    {
+      Assert(false, dealii::ExcMessage("Uninitialized LinearOperator<Range, "
+                                       "Domain>::Tvmult_add called"));
+    };
+
+    new_range_member = []() -> Range
+    {
+      Assert(false,
+      dealii::ExcMessage("Uninitialized LinearOperator<Range, "
+      "Domain>::new_range_member method called"));
+      return Range();
+    };
+
+    new_domain_member = []() -> Domain
+    {
+      Assert(false,
+      dealii::ExcMessage("Uninitialized LinearOperator<Range, "
+      "Domain>::new_domain_member method called"));
+      return Domain();
+    };
+  }
+
+  /**
+   * Create a LinearOperator object from an object @p op for which a
+   * conversion function <code>linop</code> is defined.
+   */
+  template<typename Op>
+  LinearOperator (const Op &op)
+  {
+    *this = linop(op);
+  }
+
+  /**
+   * Copy assignment operator for an object @p op for which a conversion
+   * function <code>linop</code> is defined.
+   */
+  template <typename Op>
+  LinearOperator<Range, Domain> &operator=(const Op &op)
+  {
+    return *this = linop(op);
+  }
+
+  /**
+   * Application of the LinearOperator object to a vector @p u of the
+   * Domain space giving a vector @p v of the Range space.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult;
+
+  /**
+   * Application of the LinearOperator object to a vector @p u of the
+   * Domain space. The result is added to the vector @p v.
+   */
+  std::function<const void(Range &v, const Domain &u)> vmult_add;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector @p u of
+   * the Range space giving a vector @p v of the Domain space.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult;
+
+  /**
+   * Application of the transpose LinearOperator object to a vector @p u of
+   * the Range space. The result is added to the vector @p v.
+   */
+  std::function<const void(Domain &v, const Range &u)> Tvmult_add;
+
+  /**
+   * Creates a new vector of the Range space that is completeley
+   * initialized such that it can be directly used as the destination
+   * parameter in an application of vmult.
+   */
+  std::function<Range(void)> new_range_member;
```

Comment:
Saying `(void)` hasn't been necessary in C since 1999 at least, and has never been in C++ -- might as well replace it by `()` here and below.